File: src/main/java/edu/psu/ist/plato/kaiming/ir/Printer.java
package edu.psu.ist.plato.kaiming.ir;

import java.io.OutputStream;
import java.io.PrintWriter;

import edu.psu.ist.plato.kaiming.BasicBlock;
import edu.psu.ist.plato.kaiming.util.Assert;
import edu.psu.ist.plato.kaiming.util.Tuple;

public class Printer extends PrintWriter {

    public Printer(OutputStream out) {
        super(out, true);
    }
    
    public static Printer out = new Printer(System.out);
    public static Printer err = new Printer(System.err);

    public void printConst(Const c) {
        long value = c.value();
        if (value < 0) {
            print(String.format("-0x%x", -value));
        } else {
            print(String.format("0x%x", value));
        }
    }
    
    public void printReg(Reg r) {
        print('%');
        print(r.machRegister().name());
    }
    
    public void printVar(Var v) {
        print("@(");
        print(v.name());
        print(":");
        print(v.sizeInBits());
        print(")");
    }
    
    public void printFlg(Flg f) {
        print('%');
        print(f.flag.name());
    }
    
    public void printLval(Lval lv) {
        if (lv instanceof Reg) {
            printReg((Reg)lv);
        } else if (lv instanceof Var) {
            printVar((Var)lv);
        } else if (lv instanceof Flg) {
            printFlg((Flg)lv);
        } else {
            Assert.unreachable();
        }
    }
    
    private void printBExprOperator(BExpr.Op op) {
        switch (op) {
            case ADD:    print('+'); break;
            case AND:    print('&'); break;
            case CONCAT: print(':'); break;
            case DIV:    print('/'); break;
            case MUL:    print('*'); break;
            case OR:     print('|'); break;
            case SAR:    print(">>>"); break;
            case SHL:    print("<<"); break;
            case SHR:    print(">>"); break;
            case ROR:    print("><"); break;
            case SUB:    print('-'); break;
            case XOR:    print('^'); break;
            case SEXT:   print("sext"); break;
            case UEXT:   print("uext"); break;
        }
    }
    
    private void printBExpr(BExpr e) {
        {
            boolean isLeftPrimitive = e.leftSubExpr() instanceof BExpr;
            if (isLeftPrimitive)
                print('(');
            printExpr(e.leftSubExpr());
            if (isLeftPrimitive)
                print(')');
        }
        print(' ');
        printBExprOperator(e.operator());
        print(' ');
        {
            boolean isRightPrimitive = e.rightSubExpr() instanceof BExpr;
            if (isRightPrimitive)
                print('(');
            printExpr(e.rightSubExpr());
            if (isRightPrimitive)
                print(')');
        }
    }
    
    private void printUExprOperator(UExpr.Op op) {
        switch (op) {
            case BSWAP:    print("<>");   break;
            case HIGH:     print("high"); break;
            case LOW:      print("low");  break;
            case NOT:      print("~");    break;
            case CARRY:    print("C");    break;
            case NEGATIVE: print("N");    break;
            case OVERFLOW: print("V");    break;
            case ZERO:     print("Z");    break;
        }
    }
    
    private void printUExpr(UExpr e) {
        print('(');
        printUExprOperator(e.operator());
        print(' ');
        printExpr(e.subExpr());
        print(')');
    }
    
    private void printTarget(Target l) {
        print(l.targetLabel().name());
    }
    
    public void printExpr(Expr e) {
        if (e instanceof Const) {
            printConst((Const) e);
        } else if (e instanceof Lval) {
            printLval((Lval) e);
        } else if (e instanceof BExpr) {
            printBExpr((BExpr) e);
        } else if (e instanceof UExpr) {
            printUExpr((UExpr) e);
        } else {
            Assert.unreachable();
        }
    }
    
    private void printAssignStmt(AssignStmt s) {
        printLval(s.definedLval());
        if (s.isPartialAssignment()) {
            Tuple<Integer, Integer> range = s.rangeOfAssignment();
            print('<');
            print(range.first);
            print(',');
            print(range.second);
            print('>');
        }
        print(" = ");
        printExpr(s.usedRval());
        print(";");
    }
    
    private void printCallStmt(CallStmt s) {
        print("call ");
        printExpr(s.target());
        print(";");
    }
    
    private void printCmpStmt(CmpStmt s) {
        print("cmp ");
        printExpr(s.comparedFirst());
        print(", ");
        printExpr(s.comparedSecond());
        print(";");
    }
    
    private void printJmpStmt(JmpStmt s) {
        print("jmp");
        if (s.isConditional()) {
            print('[');
            s.dependentFlags().forEach(f -> print(f.toString() + ", "));
            print(']');
        }
        print(" ");
        if (s.hasResolvedTarget())
            printTarget(s.resolvedTarget());
        else
            printExpr(s.target());
        print(";");
    }
    
    private void printLdStmt(LdStmt s) {
        printLval(s.definedLval());
        print(" <- [ ");
        printExpr(s.loadFrom());
        print(" ];");
    }

    private void printStStmt(StStmt s) {
        printExpr(s.storedExpr());
        print(" -> [ ");
        printExpr(s.storeTo());
        print(" ];");
    }
    
    public void printSelectStmt(SelStmt s) {
        printLval(s.definedLval());
        print(" = ");
        printExpr(s.condition());
        print(" ? ");
        printExpr(s.truevalue());
        print(" : ");
        printExpr(s.falsevalue());
    }
    
    public void printStmt(Stmt s) {
        switch (s.kind()) {
            case ASSIGN:
                printAssignStmt((AssignStmt) s);
                break;
            case CALL:
                printCallStmt((CallStmt) s);
                break;
            case CMP:
                printCmpStmt((CmpStmt) s);
                break;
            case JMP:
                printJmpStmt((JmpStmt) s);
                break;
            case LD:
                printLdStmt((LdStmt) s);
                break;
            case RET:
                print("return;");
                break;
            case SETF:
                print("setf;");
                break;
            case ST:
                printStStmt((StStmt) s);
                break;
            case SELECT:
                printSelectStmt((SelStmt) s);
        }
    }
    
    public void printBasicBlock(BasicBlock<Stmt> bb) {
        println(bb.label() + ":");
        bb.forEach(s -> { print('\t'); printStmt(s); println(); });
    }
    
    public void printIndentedBasicBlock(BasicBlock<Stmt> bb) {
        print('\t');
        println(bb.label() + ":");
        bb.forEach(s -> { print("\t\t"); printStmt(s); println(); });
    }
    
    public void printContextWithUDInfo(Context ctx) {
        print(ctx.name());
        println(" {");
        ctx.cfg().forEach(bb -> {
            print(bb.label());
            println(":");
            bb.forEach(s -> {
                print('\t');
                printStmt(s);
                print("\t# ");
                println(s.index());
                s.usedLvals().forEach(lv -> {
                    print("\t\t# ");
                    printLval(lv);
                    print(" -- ");
                    s.searchDefFor(lv).forEach(def -> {
                        print(def.index());
                        print(", ");
                    });
                    println();
                });
            });
        });
        println('}');
    }
    
    public void printContext(Context ctx) {
        print(ctx.name());
        println(" {");
        ctx.cfg().forEach(bb -> printBasicBlock(bb));
        println('}');
    }
}

File: src/main/scala/edu/psu/ist/plato/kaiming/ir/Printer.scala
package edu.psu.ist.plato.kaiming.ir

import java.io.PrintStream
import java.io.OutputStream

import edu.psu.ist.plato.kaiming.BBlock
import edu.psu.ist.plato.kaiming.Arch.KaiMing

object Printer {
  
  val out = new Printer(Console.out)
  val err = new Printer(Console.err)
  
  private val endOfStmt = ';' 
  
}

final class Printer(ps: OutputStream) extends PrintStream(ps) {
  
  private def EOS = print(Printer.endOfStmt)
  
  def printConst(c: Const) {
    if (c.value < 0)
      print("-0x%x".format(-c.value))
    else
      print("0x%x".format(c.value))
  }
  
  def printReg(r: Reg) {
    print('%')
    print(r.mreg.name)
  }
  
  def printVar(v: Var) {
    print("@(")
    print(v.name)
    print(":")
    print(v.sizeInBits)
    print(')')
  }
  
  def printFlg(f: Flg) {
    print('%')
    print(f.mflag.name)
  }
  
  def printLval(lv: Lval) {
    lv match {
      case r: Reg => printReg(r)
      case v: Var => printVar(v)
      case f: Flg => printFlg(f)
    }
  }
  
  def printBExpr(e: BExpr) {
    e.leftSub match {
      case be: BExpr =>
        print('(')
        printBExpr(be)
        print(')')
      case _ => printExpr(e.leftSub)
    }
    print(' ')
    e match {
      case _: Add    => print('+')
      case _: And    => print('&')
      case _: Concat => print(':')
      case _: Div    => print('/')
      case _: Mul    => print('*')
      case _: Or     => print('|')
      case _: Sar    => print(">>>")
      case _: Shl    => print("<<")
      case _: Shr    => print(">>")
      case _: Ror    => print("><")
      case _: Sub    => print('-')
      case _: Xor    => print('^')
      case _: SExt   => print("sext")
      case _: UExt   => print("uext")
    }
    print(' ')
    e.rightSub match {
      case be: BExpr =>
        print('(')
        printBExpr(be)
        print(')')
      case _ => printExpr(e.rightSub)
    }
  }
  
  def printUExpr(e: UExpr) {
    e match {
      case _: BSwap => print("<>")
      case _: High => print("high")
      case _: Low => print("low")
      case _: Not => print("~")
    }
    e.sub match {
      case be: BExpr =>
        print('(')
        printBExpr(be)
        print(')')
      case _ => printExpr(e.sub)
    }
  }
  
  def printExpr(e: Expr) {
    e match {
      case c: Const  => printConst(c)
      case lv: Lval  => printLval(lv)
      case be: BExpr => printBExpr(be)
      case ue: UExpr => printUExpr(ue)
    }
  }
  
  def printAssignStmt(s: AssignStmt) {
    printLval(s.definedLval)
    if (s.isPartial) {
      print('<')
      print(s.range._1)
      print(',')
      print(s.range._2)
      print('>')
    }
    print(" = ")
    printExpr(s.usedRval)
    EOS
  }
  
  def printCallStmt(s: CallStmt) {
    print("call ")
    printExpr(s.target)
    EOS
  }
  
  def printJmpStmt(s: JmpStmt) {
    print("jmp")
    if (s.isConditional) {
      print('[')
      print(s.dependentFlags.map { f => f.name }.mkString(","))
      print(']')
    }
    print(" ")
    s.relocatedTarget match {
      case Some(bb) => print(bb.label.name)
      case None => printExpr(s.target)
    }
    EOS
  }
  
  def printLdStmt(s: LdStmt) {
    printLval(s.definedLval)
    print(" <- [ ")
    printExpr(s.loadFrom)
    print(" ]")
    EOS
  }
  
  def printStStmt(s: StStmt) {
    printExpr(s.storedExpr)
    print(" -> [ ")
    printExpr(s.storeTo)
    print(" ]")
    EOS
  }
  
  def printSelStmt(s: SelStmt) {
    printLval(s.definedLval)
    print(" = ")
    printExpr(s.condition)
    print(" ? ")
    printExpr(s.trueValue)
    print(" : ")
    printExpr(s.falseValue)
    EOS
  }
  
  def printSetFlgStmt(s: SetFlgStmt) {
    printFlg(s.definedLval)
    print(" <~ ")
    print(s.extractor.entryName)
    print(' ')
    printExpr(s.usedRval)
    EOS
  }
  
  def printRetStmt(s: RetStmt) {
    print("ret")
    EOS
  }
  
  def printStmt(s: Stmt) {
    s match {
      case s: AssignStmt => printAssignStmt(s)
      case s: CallStmt => printCallStmt(s)
      case s: JmpStmt => printJmpStmt(s)
      case s: LdStmt => printLdStmt(s)
      case s: StStmt => printStStmt(s)
      case s: SelStmt => printSelStmt(s)
      case s: SetFlgStmt => printSetFlgStmt(s)
      case s: RetStmt => printRetStmt(s)
    }
  }
  
  def printBasicBlock(bb: BBlock[KaiMing]) {
    println(bb.label)
    bb.foreach { s => { printStmt(s); println() } }
  }
  
  def printIndentedBasicBlock(bb: BBlock[KaiMing]) {
    println(bb.label)
    bb.foreach { s => { print("\t"); printStmt(s); println() } }
  }
  
  def printContext(ctx: Context) {
    print(ctx.name);
    println(" {")
    ctx.cfg.blocks.foreach { bb => printIndentedBasicBlock(bb) }
    println('}')
  }
  
  def printContextWithUDInfo(ctx: Context) {
    print(ctx.name)
    println(" {")
    for (bb <- ctx.cfg) {
      println(bb.label)
      for (s <- bb) {
        print('\t')
        printStmt(s)
        print("\t#")
        print(s.host.index.toHexString)
        print(' ')
        println(s.index)
        ctx.definitionFor(s) match {
          case None => printStmt(s); println;
          case _ =>
        }
        for ((lv, defs) <- ctx.definitionFor(s).get) {
          print("\t\t# ")
          printLval(lv)
          print(" -- ")
          for (definition <- defs) {
            definition match {
              case Context.Def(d) => print(d.index)
              case Context.Init => print(Context.Init)
            }
            print(',')
          }
          println()
        }
      }
    }
    println('}')
  }
  
}